File: sdk/web/src/quality.ts
/**
 * Network-quality model shared with the server and the native SDKs (`aurix_common::types::quality`):
 * a simplified E-model rating `R`, a MOS estimate and 1–5 bars. Loss is always a
 * percentage (`0..=100`).
 */

/**
 * Transmission rating `R` in `0..=100`. Jitter counts double because the jitter buffer turns it
 * into delay; each percent of loss costs 2.5 points. Non-finite inputs are treated as
 * "unknown" (0 for delay, 100 for loss).
 */
export function rFactor(rttMs: number, jitterMs: number, lossPercent: number): number {
  const rtt = Number.isFinite(rttMs) ? Math.max(0, rttMs) : 0;
  const jitter = Number.isFinite(jitterMs) ? Math.max(0, jitterMs) : 0;
  const loss = Number.isFinite(lossPercent) ? clamp(lossPercent, 0, 100) : 100;
  const effectiveLatency = rtt + jitter * 2 + 10;
  const r = effectiveLatency < 160 ? 93.2 - effectiveLatency / 40 : 93.2 - (effectiveLatency - 120) / 10;
  return clamp(r - loss * 2.5, 0, 100);
}

/** Mean opinion score `1.0..=4.5` for a rating `R`. */
export function mosFromR(r: number): number {
  const x = clamp(r, 0, 100);
  return 1 + 0.035 * x + x * (x - 60) * (100 - x) * 7e-6;
}

/** `R ≥ 80` → 5 bars, `≥ 70` → 4, `≥ 60` → 3, `≥ 50` → 2, else 1. */
export function barsFromR(r: number): 1 | 2 | 3 | 4 | 5 {
  if (r >= 80) return 5;
  if (r >= 70) return 4;
  if (r >= 60) return 3;
  if (r >= 50) return 2;
  return 1;
}

/** Loss percentage from cumulative counters; `0` when nothing was expected. */
export function lossPercent(lost: number, received: number): number {
  const expected = lost + received;
  if (!(expected > 0)) return 0;
  return clamp((lost * 100) / expected, 0, 100);
}

function clamp(v: number, lo: number, hi: number): number {
  return Math.min(hi, Math.max(lo, v));
}

/** Server-side view of the connection (`NetworkQuality` control message). */
export interface NetworkQuality {
  /** 1 (unusable) … 5 (excellent), the worse of the two directions. */
  bars: number;
  rFactor: number;
  mos: number;
  /** Round trip as measured by this client and reported to the server. */
  rttMs: number;
  downlinkJitterMs: number;
  downlinkLossPercent: number;
  /** Inter-arrival jitter of this client's audio at the server (RFC 3550). */
  uplinkJitterMs: number;
  /** Sequence gaps in this client's packets over the last report interval. */
  uplinkLossPercent: number;
  uplinkBitrateKbps: number;
  uplinkPacketsReceived: number;
  uplinkPacketsLost: number;
}

export interface NetworkQualityWire {
  bars: number;
  r_factor: number;
  mos: number;
  rtt_ms: number;
  downlink_jitter_ms: number;
  downlink_loss_percent: number;
  uplink_jitter_ms: number;
  uplink_loss_percent: number;
  uplink_bitrate_kbps: number;
  uplink_packets_received: number;
  uplink_packets_lost: number;
}

export function networkQualityFromWire(w: NetworkQualityWire): NetworkQuality {
  return {
    bars: w.bars,
    rFactor: w.r_factor,
    mos: w.mos,
    rttMs: w.rtt_ms,
    downlinkJitterMs: w.downlink_jitter_ms,
    downlinkLossPercent: w.downlink_loss_percent,
    uplinkJitterMs: w.uplink_jitter_ms,
    uplinkLossPercent: w.uplink_loss_percent,
    uplinkBitrateKbps: w.uplink_bitrate_kbps,
    uplinkPacketsReceived: w.uplink_packets_received,
    uplinkPacketsLost: w.uplink_packets_lost,
  };
}

/**
 * One client-side statistics snapshot. Counters are cumulative for the peer connection;
 * `lossPercent`, `rFactor`, `mos` and `bars` describe the last `getStats()` period.
 */
export interface ClientStats {
  /** Application-level round trip from `Ping`/`Pong` (ms), with session min/avg/max. */
  rttMs: number;
  rttMinMs: number;
  rttAvgMs: number;
  rttMaxMs: number;
  /** ICE round trip from `candidate-pair.currentRoundTripTime` (ms), 0 if unknown. */
  iceRttMs: number;
  /** Inbound (downlink) audio, RFC 3550 jitter in ms. */
  jitterMs: number;
  /** Downlink loss over the last period, `0..=100`. */
  lossPercent: number;
  packetsReceived: number;
  packetsLost: number;
  bytesReceived: number;
  /** Frames the decoder concealed (PLC) and packets discarded by the jitter buffer. */
  concealedSamples: number;
  packetsDiscarded: number;
  /** Jitter buffer target/actual delay in ms (browser-reported averages). */
  jitterBufferDelayMs: number;
  /** Outbound (uplink) audio. */
  packetsSent: number;
  bytesSent: number;
  /** Uplink loss as reported by the SFU's receiver reports, `0..=100`. */
  remoteLossPercent: number;
  remoteJitterMs: number;
  /** Client-measured downlink quality. */
  rFactor: number;
  mos: number;
  bars: 1 | 2 | 3 | 4 | 5;
  /** Latest server-side quality (both directions), when received. */
  server?: NetworkQuality;
}

/** Per-period loss from cumulative `packetsLost`/`packetsReceived` counters. */
export class LossWindow {
  private prevLost = 0;
  private prevReceived = 0;
  /** Loss of the last period, `0..=100`. */
  lossPercent = 0;

  /** Advance with the current cumulative counters; returns the period loss. */
  advance(lost: number, received: number): number {
    const dLost = Math.max(0, lost - this.prevLost);
    const dRecv = Math.max(0, received - this.prevReceived);
    this.prevLost = lost;
    this.prevReceived = received;
    this.lossPercent = lossPercent(dLost, dRecv);
    return this.lossPercent;
  }

  reset(): void {
    this.prevLost = 0;
    this.prevReceived = 0;
    this.lossPercent = 0;
  }
}

/**
 * The subset of `RTCStatsReport` entries the snapshot is built from, as plain structural
 * types so the assembly can run (and be tested) outside a browser.
 */
export interface RtcStatsInput {
  inbound?: {
    jitter?: number;
    packetsReceived?: number;
    packetsLost?: number;
    bytesReceived?: number;
    concealedSamples?: number;
    packetsDiscarded?: number;
    jitterBufferDelay?: number;
    jitterBufferEmittedCount?: number;
  };
  outbound?: { packetsSent?: number; bytesSent?: number };
  remoteInbound?: { jitter?: number; packetsLost?: number; fractionLost?: number };
  /** `currentRoundTripTime` (seconds) of the selected ICE candidate pair. */
  iceRttSeconds?: number;
}

export interface RttSnapshot {
  last: number;
  min: number;
  avg: number;
  max: number;
}

/**
 * Build a `ClientStats` snapshot from WebRTC stats. `window` is advanced with the cumulative
 * inbound counters, so the loss percentage covers the period since the previous call. The
 * rating uses the application RTT when known, otherwise the ICE RTT.
 */
export function assembleClientStats(
  rtt: RttSnapshot,
  input: RtcStatsInput,
  window: LossWindow,
  server: NetworkQuality | undefined,
): ClientStats {
  const out: ClientStats = {
    rttMs: rtt.last,
    rttMinMs: rtt.min,
    rttAvgMs: rtt.avg,
    rttMaxMs: rtt.max,
    iceRttMs: input.iceRttSeconds !== undefined && Number.isFinite(input.iceRttSeconds) ? input.iceRttSeconds * 1000 : 0,
    jitterMs: 0,
    lossPercent: window.lossPercent,
    packetsReceived: 0,
    packetsLost: 0,
    bytesReceived: 0,
    concealedSamples: 0,
    packetsDiscarded: 0,
    jitterBufferDelayMs: 0,
    packetsSent: 0,
    bytesSent: 0,
    remoteLossPercent: 0,
    remoteJitterMs: 0,
    rFactor: 0,
    mos: 0,
    bars: 1,
    ...(server ? { server } : {}),
  };
  const i = input.inbound;
  if (i) {
    out.jitterMs = (i.jitter ?? 0) * 1000;
    out.packetsReceived = i.packetsReceived ?? 0;
    out.packetsLost = Math.max(0, i.packetsLost ?? 0);
    out.bytesReceived = i.bytesReceived ?? 0;
    out.concealedSamples = i.concealedSamples ?? 0;
    out.packetsDiscarded = i.packetsDiscarded ?? 0;
    const emitted = i.jitterBufferEmittedCount ?? 0;
    out.jitterBufferDelayMs = emitted > 0 ? ((i.jitterBufferDelay ?? 0) * 1000) / emitted : 0;
    out.lossPercent = window.advance(out.packetsLost, out.packetsReceived);
  }
  if (input.outbound) {
    out.packetsSent = input.outbound.packetsSent ?? 0;
    out.bytesSent = input.outbound.bytesSent ?? 0;
  }
  const r = input.remoteInbound;
  if (r) {
    out.remoteJitterMs = (r.jitter ?? 0) * 1000;
    out.remoteLossPercent =
      r.fractionLost !== undefined && Number.isFinite(r.fractionLost)
        ? clamp(r.fractionLost * 100, 0, 100)
        : lossPercent(Math.max(0, r.packetsLost ?? 0), out.packetsSent);
  }
  const rating = rFactor(out.rttMs > 0 ? out.rttMs : out.iceRttMs, out.jitterMs, out.lossPercent);
  out.rFactor = rating;
  out.mos = mosFromR(rating);
  out.bars = barsFromR(rating);
  return out;
}

/** Running min/avg/max of RTT samples over a session. */
export class RttTracker {
  last = 0;
  min = 0;
  max = 0;
  avg = 0;
  samples = 0;

  record(rttMs: number): void {
    if (!Number.isFinite(rttMs) || rttMs < 0) return;
    this.last = rttMs;
    if (this.samples === 0) {
      this.min = this.max = this.avg = rttMs;
    } else {
      this.min = Math.min(this.min, rttMs);
      this.max = Math.max(this.max, rttMs);
      this.avg += (rttMs - this.avg) / (this.samples + 1);
    }
    this.samples += 1;
  }

  reset(): void {
    this.last = this.min = this.max = this.avg = 0;
    this.samples = 0;
  }
}

File: sdk/unity/Runtime/Quality.cs
using System;
using System.Collections.Generic;
using Aurix.Protocol;

namespace Aurix
{
    /// <summary>
    /// Network-quality model shared with the server and the other SDKs
    /// (<c>aurix_common::types::quality</c>): a simplified E-model rating <c>R</c>, a MOS estimate and
    /// 1–5 bars. Loss is always a percentage (<c>0..100</c>).
    /// </summary>
    public static class QualityModel
    {
        /// <summary>
        /// Transmission rating <c>R</c> in <c>0..100</c>. Jitter counts double because the jitter buffer
        /// turns it into delay; each percent of loss costs 2.5 points. Non-finite inputs are treated as
        /// unknown (0 for delay, 100 for loss).
        /// </summary>
        public static float RFactor(float rttMs, float jitterMs, float lossPercent)
        {
            float rtt = IsFinite(rttMs) ? Math.Max(0f, rttMs) : 0f;
            float jitter = IsFinite(jitterMs) ? Math.Max(0f, jitterMs) : 0f;
            float loss = IsFinite(lossPercent) ? Clamp(lossPercent, 0f, 100f) : 100f;
            float effectiveLatency = rtt + jitter * 2f + 10f;
            float r = effectiveLatency < 160f ? 93.2f - effectiveLatency / 40f : 93.2f - (effectiveLatency - 120f) / 10f;
            return Clamp(r - loss * 2.5f, 0f, 100f);
        }

        /// <summary>Mean opinion score <c>1.0..4.5</c> for a rating <c>R</c>.</summary>
        public static float MosFromR(float r)
        {
            float x = Clamp(r, 0f, 100f);
            return 1f + 0.035f * x + x * (x - 60f) * (100f - x) * 7e-6f;
        }

        /// <summary><c>R ≥ 80</c> → 5 bars, <c>≥ 70</c> → 4, <c>≥ 60</c> → 3, <c>≥ 50</c> → 2, else 1.</summary>
        public static int BarsFromR(float r)
        {
            if (r >= 80f) return 5;
            if (r >= 70f) return 4;
            if (r >= 60f) return 3;
            if (r >= 50f) return 2;
            return 1;
        }

        /// <summary>Loss percentage from counters; 0 when nothing was expected.</summary>
        public static float LossPercent(long lost, long received)
        {
            long expected = lost + received;
            if (expected <= 0) return 0f;
            return Clamp(lost * 100f / expected, 0f, 100f);
        }

        private static bool IsFinite(float v) => !float.IsNaN(v) && !float.IsInfinity(v);
        private static float Clamp(float v, float lo, float hi) => v < lo ? lo : v > hi ? hi : v;
    }

    /// <summary>Per-period loss from cumulative lost/received counters.</summary>
    public sealed class LossWindow
    {
        private long _prevLost, _prevReceived;

        /// <summary>Loss of the last period, <c>0..100</c>.</summary>
        public float LossPercent { get; private set; }

        /// <summary>Advance with the current cumulative counters; returns the loss of the period since the previous call.</summary>
        public float Advance(long lost, long received)
        {
            long dLost = Math.Max(0, lost - _prevLost);
            long dRecv = Math.Max(0, received - _prevReceived);
            _prevLost = lost;
            _prevReceived = received;
            LossPercent = QualityModel.LossPercent(dLost, dRecv);
            return LossPercent;
        }

        public void Reset()
        {
            _prevLost = _prevReceived = 0;
            LossPercent = 0f;
        }
    }

    /// <summary>
    /// The server's view of this connection (<c>NetworkQuality</c> control message): the downlink as
    /// reported by this client merged with what the SFU measures on the uplink; <see cref="Bars"/> is
    /// the worse of the two directions.
    /// </summary>
    public struct NetworkQuality
    {
        /// <summary>1 (unusable) … 5 (excellent).</summary>
        public int Bars;
        public float RFactor;
        public float Mos;
        /// <summary>Round trip as measured and reported by this client.</summary>
        public float RttMs;
        public float DownlinkJitterMs;
        public float DownlinkLossPercent;
        /// <summary>Inter-arrival jitter of this client's audio at the server (RFC 3550).</summary>
        public float UplinkJitterMs;
        /// <summary>Sequence gaps in this client's packets over the last report interval.</summary>
        public float UplinkLossPercent;
        public uint UplinkBitrateKbps;
        public long UplinkPacketsReceived;
        public long UplinkPacketsLost;

        /// <summary>Typed view of a <c>NetworkQuality</c> payload (<c>data.quality</c>).</summary>
        public static NetworkQuality? FromMessage(ControlMessage m)
        {
            if (m.Data == null || !m.Data.TryGetValue("quality", out var qv)) return null;
            var q = MiniJson.AsObject(qv);
            if (q == null) return null;
            return FromObject(q);
        }

        public static NetworkQuality FromObject(Dictionary<string, object> q) => new NetworkQuality
        {
            Bars = (int)MiniJson.GetNumber(q, "bars"),
            RFactor = (float)MiniJson.GetNumber(q, "r_factor"),
            Mos = (float)MiniJson.GetNumber(q, "mos"),
            RttMs = (float)MiniJson.GetNumber(q, "rtt_ms"),
            DownlinkJitterMs = (float)MiniJson.GetNumber(q, "downlink_jitter_ms"),
            DownlinkLossPercent = (float)MiniJson.GetNumber(q, "downlink_loss_percent"),
            UplinkJitterMs = (float)MiniJson.GetNumber(q, "uplink_jitter_ms"),
            UplinkLossPercent = (float)MiniJson.GetNumber(q, "uplink_loss_percent"),
            UplinkBitrateKbps = (uint)MiniJson.GetNumber(q, "uplink_bitrate_kbps"),
            UplinkPacketsReceived = (long)MiniJson.GetNumber(q, "uplink_packets_received"),
            UplinkPacketsLost = (long)MiniJson.GetNumber(q, "uplink_packets_lost"),
        };
    }

    /// <summary>
    /// One statistics snapshot of an <see cref="AurixVoiceClient"/>. Packet, byte and frame counters
    /// are cumulative for the current media transport; <see cref="LossPercent"/>, <see cref="RFactor"/>,
    /// <see cref="Mos"/> and <see cref="Bars"/> describe the last quality period.
    /// </summary>
    public struct VoiceStats
    {
        public VoiceConnectionState State;
        // Transport (native AURX over UDP).
        public long PacketsSent;
        public long BytesSent;
        public long PacketsReceived;
        public long BytesReceived;
        /// <summary>Downlink packets that failed authentication/decryption.</summary>
        public long BadAuth;
        /// <summary>Downlink packets rejected by the replay window.</summary>
        public long Replayed;
        public long HeartbeatsLost;
        /// <summary>Heartbeat RTT (media path), current and session min/avg/max.</summary>
        public float RttMs;
        public float RttMinMs;
        public float RttAvgMs;
        public float RttMaxMs;
        /// <summary>WebSocket <c>Ping</c>/<c>Pong</c> RTT.</summary>
        public float ControlRttMs;
        /// <summary>RFC 3550 inter-arrival jitter of downlink audio.</summary>
        public float JitterMs;
        /// <summary>Downlink loss over the last period, <c>0..100</c> (frames lost vs. frames received).</summary>
        public float LossPercent;
        // Playout (jitter buffers / mixer), lifetime.
        public long FramesLost;
        public long FramesLate;
        public long Underruns;
        public int ActiveStreams;
        // Derived quality.
        public float RFactor;
        public float Mos;
        /// <summary>1 (unusable) … 5 (excellent).</summary>
        public int Bars;
        /// <summary>Latest server-side quality (both directions), when received.</summary>
        public NetworkQuality? Server;
    }
}
